Refactor get_reason

kellertuer opened the pull request:
This resolves #389.
I had a bit of a boring train ride today and worked through all stopping criteria, removed `.reason` fields and when necessary introduced values to store obtained values being printed when generating the string.

There are two small things left

* [x] for the Lagrange Multiplier criterion in the bundle methods the symbols should maybe be strings since they might also be `-\xi` (and not just `\xi` like now)
* [x] somehow I have to check what happens with `JuMP` their 3 checks for reasons return now empty strings and I do no directly see why

mateuszbaran commented on src/plans/stopping_criterion.jl:
```diff
@@ -1070,3 +1124,12 @@ end
 end
+
+@doc raw"""
+    get_reason(s)
```
```suggestion
    get_reason(s::AbstractManoptSolverState)
```

mateuszbaran commented on src/plans/bundle_plan.jl:
```diff
@@ -100,32 +100,56 @@ but a vector of two values is required for the `:both` mode.
+} <: StoppingCriterion
+    tolerances::A
+    values::A
+    sumbols::B
```
```suggestion
    symbols::B
```

mateuszbaran commented:
> * for the Lagrange Multiplier criterion in the bundle methods the symbols should maybe be strings since they might also be `-\xi` (and not just `\xi` like now)

That can be saved in a separate field in the criterion but leaving the string in this one place would be fine too.

mateuszbaran commented on src/solvers/Lanczos.jl:
```diff
@@ -244,22 +244,23 @@ m(X_k) \leq m(0)
-    StopWhenFirstOrderProgress(θ::Float64) = new(θ, "")
+mutable struct StopWhenFirstOrderProgress{F} <: StoppingCriterion
+    θ::F
+    at_iteration::Integer
```
```suggestion
    at_iteration::Int
```

mateuszbaran commented on src/plans/stopping_criterion.jl:
```diff
@@ -825,20 +884,26 @@ reasons.
-    reason::String
-    StopWhenAll(c::Vector{StoppingCriterion}) = new{typeof(tuple(c...))}(tuple(c...), "")
-    StopWhenAll(c...) = new{typeof(c)}(c, "")
+    at_iteration::Integer
```
```suggestion
    at_iteration::Int
```

kellertuer commented:
Oh, I missed one solver, in CMA-ES, which has quite a few stopping criteria, they always returned a reason even if they were not active.

For now I think I did rewrite them, but testing them is still missing.
We might take the chance and change the `is_active` fields in these and change them to `at_iteration` in order to provide the iteration number in the reason (at which the stopping criterion indicated to stop).

mateuszbaran commented:
> We might take the chance and change the `is_active` fields in these and change them to `at_iteration` in order to provide the iteration number in the reason (at which the stopping criterion indicated to stop).

Yes, I think changing it to `at_iteration` would make sense.

kellertuer commented:
> > We might take the chance and change the `is_active` fields in these and change them to `at_iteration` in order to provide the iteration number in the reason (at which the stopping criterion indicated to stop).
> 
> Yes, I think changing it to `at_iteration` would make sense.

Did that. Also tried to use the same tests for `get_reason` as for others and noticed that on construction, the `StopWhenEvolutionStagnates`s show method errors (not affecting tests for now, but on REPL it did); since one of the means to compute is (on construction) over an empty array.

kellertuer commented:
Hopefully also fixed the last error I mentioned above, which was mainly that the median of an empty vector does not exist, so I introduced a check and tested that as well.
